File: September/2026-09-05/98/98-valid-binary-tree.cpp
// 98-valid-binary-tree.cpp;bt-recusion;help;leetcode

/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    bool checkBST(TreeNode* root, long long minValue, long long maxValue) {
        if (root == nullptr) {
            return true;
        }

        if (root->val <= minValue || root->val >= maxValue) {
            return false;
        }
        return checkBST(root->left, minValue, root->val) && checkBST(root->right, root->val, maxValue);
    }

    bool isValidBST(TreeNode* root) {
        return checkBST(root, LLONG_MIN, LLONG_MAX);
    }
};

File: September 2026/2026-09-05/98-valid-binary-tree.py
# 98-valid-binary-tree.cpp;bt-recusion;help;leetcode

# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def isValidBST(self, root: Optional[TreeNode]) -> bool:
        previous = [None]

        def inorder(node):
            if node is None:
                return True
            if not inorder(node.left):
                return False
            if previous[0] is not None and node.val <= previous[0]:
                return False
            previous[0] = node.val

            return inorder(node.right)

        return inorder(root)